truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java

```diff
@@ -63,20 +63,32 @@ public CachedDispatchNode(
         this.next = next;
     }
 
+    protected boolean sendingFrames() {
+        return sendsFrame || readCaller != null;
+    }
+
     public void replaceSendingFrame() {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert readCaller == null;
         if (!sendsFrame) {
-            sendsFrame = true;
+            replaceSendingFrame(true, null);
         }
     }
 
     public void replaceSendingCallerFrame(CallerFrameAccess access) {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert sendsFrame == false;
         if (readCaller == null) {
-            readCaller = new ReadCallerFrameNode(access);
+            replaceSendingFrame(false, new ReadCallerFrameNode(access));
         }
     }
 
+    private void replaceSendingFrame(boolean sendsFrame, ReadCallerFrameNode readCaller) {
+        CachedDispatchNode copy = (CachedDispatchNode) copy();
```

Comment:
Christian Humer says the clean solution is to have one Assumption per InternalMethod, and invalidate that one whenever we change the "need to send the caller frame" flag. That Assumption would be checked in all CachedDispatchNodes.